Source/JavaScriptCore/runtime/WaiterListManager.cpp

```diff
@@ -0,0 +1,241 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#pragma once
+
+#include "config.h"
+#include "WaiterListManager.h"
+
+#include "JSGlobalObject.h"
+#include "JSLock.h"
+#include "ObjectConstructor.h"
+#include <wtf/DataLog.h>
+#include <wtf/NeverDestroyed.h>
+#include <wtf/RawPointer.h>
+
+namespace JSC {
+
+namespace WaiterListsManagerInternal {
+static constexpr bool verbose = false;
+}
+
+WaiterListManager& WaiterListManager::singleton()
+{
+    static LazyNeverDestroyed<WaiterListManager> manager;
+    static std::once_flag onceKey;
+    std::call_once(onceKey, [&] {
+        manager.construct();
+    });
+    return manager;
+}
+
+JSValue WaiterListManager::wait(JSGlobalObject* globalObject, VM& vm, void* ptr, AtomicsWaitValidation validation, Seconds timeout, const AtomicsWaitType type)
+{
+    // bool didPassValidation = WTF::atomicLoad(ptr) == expectedValue;
+    if (type == AtomicsWaitType::Async) {
+        JSObject* object = constructEmptyObject(globalObject->vm(), globalObject->objectStructureForObjectConstructor());
+
+        bool isAsync = false;
+        JSValue value;
+        if (validation == AtomicsWaitValidation::Fail)
+            value = vm.smallStrings.notEqualString();
+        else if (!timeout)
+            value = vm.smallStrings.timedOutString();
+        else {
+            isAsync = true;
+            JSPromise* promise = JSPromise::create(vm, globalObject->promiseStructure());
+            addAsyncWaiter(ptr, promise, timeout);
+            value = promise;
+        }
+
+        object->putDirect(vm, vm.propertyNames->async, jsBoolean(isAsync));
+        object->putDirect(vm, vm.propertyNames->value, value);
+        return object;
+    }
+
+    if (validation == AtomicsWaitValidation::Pass) {
+        Waiter waiter { &vm };
+        WaiterList* list = addSyncWaiter(ptr, &waiter);
+        const TimeWithDynamicClockType& time = MonotonicTime::now() + timeout;
```

Comment:
Let's not take a same `list->lock` twice. It is costly. We should take it only once, and do the following sequence in an atomic manner.

```
Waiter waiter { &vm };
TimeWithDynamicClockType time = MonotonicTime::now() + timeout;
WaiterList* list = findOrCreateList(ptr);
{
    Locker locker { list->lock };
    list->addLast(locker, waiter);
    while (waiter.getVM() && time.nowWithSameClock() < time)
        waiter.getCondition().waitUntil(list->lock, time.approximateWallTime());
    RELEASE_ASSERT(!waiter.getVM() || waiter.getVM() == &vm);
    bool didGetDequeued = !waiter.getVM();
    if (didGetDequeued)
        return vm.smallStrings.okString();
    list->take(locker, &waiter);
    return vm.smallStrings.timedOutString();
}
```

And we should pass `const AbstractLocker&` parameter to the function which needs to be called while taking a lock. This is WebKit's convention.
And not use `LockHolder`. Use `Locker { list->lock }`.

And I don't think we need to make it RELEASE_ASSERT while taking a lock. Let's use ASSERT.